lib/linked_list.rb

```diff
@@ -18,132 +18,262 @@ def initialize
+    # Space Complexity: O(n) - where n equals the length of the linked list
     def visit
-      raise NotImplementedError
+      return if !@head
+
+      current = @head
+      list_string = ""
+
+      until !current
+        list_string << "#{current.data} "
+        current = current.next
+      end
+
+      print list_string
+      return list_string
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n) - where n is equal to the length of the linked list
+    # Space Complexity: O(1)
     def delete(value)
-      raise NotImplementedError
+      return if !@head
+
+      previous = nil
+      current = @head
+
+      until !current
+        if current.data == value 
+          @head.data == value ? @head = current.next : previous.next = current.next
+        end
+
+        previous = current
+        current = current.next
+      end
     end
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n) - where n is equal to the length of the linked list
+    # Space Complexity: O(1)
     def reverse
-      raise NotImplementedError
+      return if !@head || !@head.next
+
+      previous = nil
+      current = @head
+
+      while current
+        subsequent = current.next
+        current.next = previous
+        previous = current
+        current = subsequent
+      end
+
+      @head = previous
     end
 
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n) - where n is equal to the length of the linked list
+    # Space Complexity: O(1)
     def find_middle_value
-      raise NotImplementedError
+      return if !@head
+
+      current = @head
+      (self.length / 2).times { current = current.next }
+
+      return current.data
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n) - where n equals the length of the linked list
+    # Space Complexity: O(1)
     def find_nth_from_end(n)
-      raise NotImplementedError
+      return if self.length <= n
+        
+      current = @head
+      (length - n - 1).times { current = current.next} 
+    
+      return current.data
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
-    # Time Complexity:  
-    # Space Complexity
+    # Time Complexity: O(n^2) - where n equals the length of the linked list
+    # Space Complexity: O(n) - where n equals the length of the linked list
     def has_cycle

```

Comment:
You're right  in that this is an algorithm with O(n<sup>2</sup>) time complexity and O(n) space complexity.  You can do it with one loop and two references, one moving twice as fast as the other.  There's a link [here](https://www.geeksforgeeks.org/detect-loop-in-a-linked-list/)